Data Race in Pipelinerun reconciler tests

The reconciler tests seem to occasionally fail with a panic. Looks like https://github.com/tektoncd/pipeline/blob/master/pkg/reconciler/timeout_handler.go#L265 is being called after the test exits. (My assumption is that the [`defer cancel()`](https://github.com/tektoncd/pipeline/blob/master/pkg/reconciler/v1alpha1/pipelinerun/pipelinerun_test.go#L741) should ensure that the timeout goroutines get cleaned up but I understand very little about how this works)

### Additional Information
` go test -run "^TestReconcilePropagateLabels" ./pkg/reconciler/v1alpha1/pipelinerun/ -count 10`

```
panic: Log in goroutine after TestReconcilePropagateLabels/without_pipelinetask_name has completed

goroutine 101 [running]:
testing.(*common).logDepth(0xc000222300, 0xc000786240, 0xbd, 0x3)
        /usr/lib/google-golang/src/testing/testing.go:634 +0x3a6
testing.(*common).log(...)
        /usr/lib/google-golang/src/testing/testing.go:614
testing.(*common).Logf(0xc000222300, 0x18b328b, 0x2, 0xc0003b4080, 0x1, 0x1)
        /usr/lib/google-golang/src/testing/testing.go:649 +0x7f
github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zaptest.testingWriter.Write(0x1bd0060, 0xc000222300, 0x0, 0xc0007a8400, 0xbe, 0x400, 0xc0007a8400, 0x1635f40, 0xc0003b4070)
        /usr/local/google/home/dibyajyoti/dev/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zaptest/logger.go:130 +0xfc
github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore.(*ioCore).Write(0xc0003d7050, 0x0, 0xbf46480bc791818a, 0x10edee7, 0x2b0e160, 0xc0004fcf00, 0x36, 0xc000358000, 0x42, 0x1, ...)
        /usr/local/google/home/dibyajyoti/dev/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore/core.go:90 +0x107
github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc000457550, 0x0, 0x0, 0x0)
        /usr/local/google/home/dibyajyoti/dev/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore/entry.go:215 +0x119
github.com/tektoncd/pipeline/vendor/go.uber.org/zap.(*SugaredLogger).log(0xc0003de388, 0x0, 0x18c895c, 0x15, 0xc000719de8, 0x1, 0x1, 0x0, 0x0, 0x0)
        /usr/local/google/home/dibyajyoti/dev/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/sugar.go:234 +0x101
github.com/tektoncd/pipeline/vendor/go.uber.org/zap.(*SugaredLogger).Infof(...)
        /usr/local/google/home/dibyajyoti/dev/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/sugar.go:138
github.com/tektoncd/pipeline/pkg/reconciler.(*TimeoutSet).setTimer(0xc000445c80, 0x1b7b400, 0xc00023e780, 0x34630b755cc, 0xc0002de3b0)
        /usr/local/google/home/dibyajyoti/dev/go/src/github.com/tektoncd/pipeline/pkg/reconciler/timeout_handler.go:265 +0x4d3
github.com/tektoncd/pipeline/pkg/reconciler.(*TimeoutSet).waitRun(0xc000445c80, 0x1b7b400, 0xc00023e780, 0x34630b8a000, 0xc0000f4640, 0xc0002de3b0)
        /usr/local/google/home/dibyajyoti/dev/go/src/github.com/tektoncd/pipeline/pkg/reconciler/timeout_handler.go:241 +0x29d
github.com/tektoncd/pipeline/pkg/reconciler.(*TimeoutSet).WaitPipelineRun(0xc000445c80, 0xc00023e780, 0xc0000f4640)
        /usr/local/google/home/dibyajyoti/dev/go/src/github.com/tektoncd/pipeline/pkg/reconciler/timeout_handler.go:227 +0x68
created by github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/pipelinerun.(*Reconciler).Reconcile
        /usr/local/google/home/dibyajyoti/dev/go/src/github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go:141 +0x1224
FAIL    github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/pipelinerun        0.221s
```

hrm, another data point, I'm able to repro.  go version 1.12.5

```
sbws@sbws18:~/dev/go/src/github.com/tektoncd/pipeline
 (master) λ go test -run "^TestReconcilePropagateLabels" ./pkg/reconciler/v1alpha1/pipelinerun/ -count 10
panic: Log in goroutine after TestReconcilePropagateLabels/with_pipelinetask_name has completed [recovered]
	panic: Log in goroutine after TestReconcilePropagateLabels/with_pipelinetask_name has completed

goroutine 35 [running]:
testing.tRunner.func1(0xc000373900)
	/usr/lib/google-golang/src/testing/testing.go:830 +0x392
panic(0x1639d20, 0xc0002b91b0)
	/usr/lib/google-golang/src/runtime/panic.go:522 +0x1b5
testing.(*common).logDepth(0xc0002c2400, 0xc0001068c0, 0xd3, 0x3)
	/usr/lib/google-golang/src/testing/testing.go:634 +0x3a6
testing.(*common).log(...)
	/usr/lib/google-golang/src/testing/testing.go:614
testing.(*common).Logf(0xc0002c2400, 0x18bee2d, 0x2, 0xc0002b91a0, 0x1, 0x1)
	/usr/lib/google-golang/src/testing/testing.go:649 +0x7f
github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zaptest.testingWriter.Write(0x1bdaa00, 0xc0002c2400, 0x0, 0xc000167c00, 0xd4, 0x400, 0xc000167c00, 0x1639d20, 0xc0002b9190)
	/usr/local/google/home/sbws/dev/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zaptest/logger.go:130 +0xfc
github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore.(*ioCore).Write(0xc0001c2420, 0xff, 0xbf4831fee116d9cc, 0x189ff75, 0x2b29400, 0xc00004f040, 0x47, 0xc0005dd460, 0x1a, 0x1, ...)
	/usr/local/google/home/sbws/dev/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore/core.go:90 +0x107
github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc000667ef0, 0x0, 0x0, 0x0)
	/usr/local/google/home/sbws/dev/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore/entry.go:215 +0x119
github.com/tektoncd/pipeline/vendor/go.uber.org/zap.(*SugaredLogger).log(0xc0000105f0, 0xc0000b57ff, 0x0, 0x0, 0xc0000b5750, 0x1, 0x1, 0x0, 0x0, 0x0)
	/usr/local/google/home/sbws/dev/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/sugar.go:234 +0x101
github.com/tektoncd/pipeline/vendor/go.uber.org/zap.(*SugaredLogger).Debug(...)
	/usr/local/google/home/sbws/dev/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/sugar.go:97
github.com/tektoncd/pipeline/pkg/reconciler.NewBase(0x1c0fb40, 0xc0000d88c0, 0x1baf900, 0xc0000d8820, 0x0, 0x0, 0x1ba7600, 0xc0000c2af0, 0xc000010038, 0x0, ...)
	/usr/local/google/home/sbws/dev/go/src/github.com/tektoncd/pipeline/pkg/reconciler/reconciler.go:95 +0x3db
github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/pipelinerun.NewController(0x1bcbac0, 0xc000089b80, 0x1ba7600, 0xc0000c2af0, 0xc0002b90e0)
	/usr/local/google/home/sbws/dev/go/src/github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/pipelinerun/controller.go:68 +0x3af
github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/pipelinerun.getPipelineRunController(0xc000373900, 0xc000079e38, 0x1, 0x1, 0xc000079e40, 0x1, 0x1, 0x0, 0x0, 0x0, ...)
	/usr/local/google/home/sbws/dev/go/src/github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/pipelinerun/pipelinerun_test.go:54 +0x173
github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/pipelinerun.TestReconcilePropagateLabels.func1(0xc000373900)
	/usr/local/google/home/sbws/dev/go/src/github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/pipelinerun/pipelinerun_test.go:748 +0x410
testing.tRunner(0xc000373900, 0xc0002b8e70)
	/usr/lib/google-golang/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
	/usr/lib/google-golang/src/testing/testing.go:916 +0x35a
FAIL	github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/pipelinerun	0.218s
```

Tried with different settings of GOMAXPROCS but no change to the result, it panics each time I run it.

I've done some testing around this one.
Only looking at `pipelinerun_test.go`, it seems that the following tests generate issues:

```
-func TestReconcile_InvalidPipelineRuns(t *testing.T) {
-func TestReconcilePropagateLabels(t *testing.T) {
-func TestReconcileWithConditionChecks(t *testing.T) {
-func TestReconcileWithFailingConditionChecks(t *testing.T) {```

If I enable `TestReconcileWithFailingConditionChecks` or `TestReconcileWithConditionChecks` I get a panic but no data race:
```
$ go test -race  ./pkg/reconciler/pipelinerun/
E0911 16:40:28.050423   72124 event.go:259] Could not construct reference to: '&v1alpha1.PipelineRun{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"test-pipeline-run-with-timeout", GenerateName:"", Namespace:"foo", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"tekton.dev/pipeline":"test-pipeline"}, Annotations:map[string]string{}, OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:v1alpha1.PipelineRunSpec{PipelineRef:v1alpha1.PipelineRef{Name:"test-pipeline", APIVersion:""}, Resources:[]v1alpha1.PipelineResourceBinding(nil), Params:[]v1alpha1.Param(nil), ServiceAccount:"test-sa", ServiceAccounts:[]v1alpha1.PipelineRunSpecServiceAccount(nil), Results:(*v1alpha1.Results)(nil), Status:"", Timeout:(*v1.Duration)(0xc00011b4c0), PodTemplate:v1alpha1.PodTemplate{NodeSelector:map[string]string(nil), Tolerations:[]v1.Toleration(nil), Affinity:(*v1.Affinity)(nil), SecurityContext:(*v1.PodSecurityContext)(nil), Volumes:[]v1.Volume(nil)}, NodeSelector:map[string]string(nil), Tolerations:[]v1.Toleration(nil), Affinity:(*v1.Affinity)(nil)}, Status:v1alpha1.PipelineRunStatus{Status:v1beta1.Status{ObservedGeneration:0, Conditions:v1beta1.Conditions{apis.Condition{Type:"Succeeded", Status:"False", Severity:"", LastTransitionTime:apis.VolatileTime{Inner:v1.Time{Time:time.Time{wall:0xbf56633b0300ff48, ext:71671574, loc:(*time.Location)(0x4314120)}}}, Reason:"PipelineRunTimeout", Message:"PipelineRun \"test-pipeline-run-with-timeout\" failed to finish within \"&Duration{Duration:12h0m0s,}\""}}}, Results:(*v1alpha1.Results)(nil), StartTime:(*v1.Time)(0xc00000e720), CompletionTime:(*v1.Time)(nil), TaskRuns:map[string]*v1alpha1.PipelineRunTaskRunStatus{}}}' due to: 'selfLink was empty, can't make reference'. Will not report event: 'Warning' 'Failed' 'PipelineRun "test-pipeline-run-with-timeout" failed to finish within "&Duration{Duration:12h0m0s,}"'
E0911 16:40:28.061250   72124 event.go:259] Could not construct reference to: '&v1alpha1.PipelineRun{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"test-pipeline-retry-run-with-timeout", GenerateName:"", Namespace:"foo", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"tekton.dev/pipeline":"test-pipeline-retry"}, Annotations:map[string]string{}, OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:v1alpha1.PipelineRunSpec{PipelineRef:v1alpha1.PipelineRef{Name:"test-pipeline-retry", APIVersion:""}, Resources:[]v1alpha1.PipelineResourceBinding(nil), Params:[]v1alpha1.Param(nil), ServiceAccount:"test-sa", ServiceAccounts:[]v1alpha1.PipelineRunSpecServiceAccount(nil), Results:(*v1alpha1.Results)(nil), Status:"", Timeout:(*v1.Duration)(0xc00011a2a0), PodTemplate:v1alpha1.PodTemplate{NodeSelector:map[string]string(nil), Tolerations:[]v1.Toleration(nil), Affinity:(*v1.Affinity)(nil), SecurityContext:(*v1.PodSecurityContext)(nil), Volumes:[]v1.Volume(nil)}, NodeSelector:map[string]string(nil), Tolerations:[]v1.Toleration(nil), Affinity:(*v1.Affinity)(nil)}, Status:v1alpha1.PipelineRunStatus{Status:v1beta1.Status{ObservedGeneration:0, Conditions:v1beta1.Conditions{apis.Condition{Type:"Succeeded", Status:"False", Severity:"", LastTransitionTime:apis.VolatileTime{Inner:v1.Time{Time:time.Time{wall:0xbf56633b03a66af0, ext:82513790, loc:(*time.Location)(0x4314120)}}}, Reason:"PipelineRunTimeout", Message:"PipelineRun \"test-pipeline-retry-run-with-timeout\" failed to finish within \"&Duration{Duration:12h0m0s,}\""}}}, Results:(*v1alpha1.Results)(nil), StartTime:(*v1.Time)(0xc0006a63e0), CompletionTime:(*v1.Time)(nil), TaskRuns:map[string]*v1alpha1.PipelineRunTaskRunStatus{"hello-world-1":(*v1alpha1.PipelineRunTaskRunStatus)(0xc0006a6400)}}}' due to: 'selfLink was empty, can't make reference'. Will not report event: 'Warning' 'Failed' 'PipelineRun "test-pipeline-retry-run-with-timeout" failed to finish within "&Duration{Duration:12h0m0s,}"'
E0911 16:40:28.063814   72124 event.go:259] Could not construct reference to: '&v1alpha1.PipelineRun{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"test-pipeline-retry-run-with-timeout", GenerateName:"", Namespace:"foo", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"tekton.dev/pipeline":"test-pipeline-retry"}, Annotations:map[string]string{}, OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:v1alpha1.PipelineRunSpec{PipelineRef:v1alpha1.PipelineRef{Name:"test-pipeline-retry", APIVersion:""}, Resources:[]v1alpha1.PipelineResourceBinding(nil), Params:[]v1alpha1.Param(nil), ServiceAccount:"test-sa", ServiceAccounts:[]v1alpha1.PipelineRunSpecServiceAccount(nil), Results:(*v1alpha1.Results)(nil), Status:"", Timeout:(*v1.Duration)(0xc00011b2b8), PodTemplate:v1alpha1.PodTemplate{NodeSelector:map[string]string(nil), Tolerations:[]v1.Toleration(nil), Affinity:(*v1.Affinity)(nil), SecurityContext:(*v1.PodSecurityContext)(nil), Volumes:[]v1.Volume(nil)}, NodeSelector:map[string]string(nil), Tolerations:[]v1.Toleration(nil), Affinity:(*v1.Affinity)(nil)}, Status:v1alpha1.PipelineRunStatus{Status:v1beta1.Status{ObservedGeneration:0, Conditions:v1beta1.Conditions{apis.Condition{Type:"Succeeded", Status:"False", Severity:"", LastTransitionTime:apis.VolatileTime{Inner:v1.Time{Time:time.Time{wall:0xbf56633b03cd9260, ext:85079769, loc:(*time.Location)(0x4314120)}}}, Reason:"PipelineRunTimeout", Message:"PipelineRun \"test-pipeline-retry-run-with-timeout\" failed to finish within \"&Duration{Duration:12h0m0s,}\""}}}, Results:(*v1alpha1.Results)(nil), StartTime:(*v1.Time)(0xc0006a76e0), CompletionTime:(*v1.Time)(nil), TaskRuns:map[string]*v1alpha1.PipelineRunTaskRunStatus{"hello-world-1":(*v1alpha1.PipelineRunTaskRunStatus)(0xc0006a7700)}}}' due to: 'selfLink was empty, can't make reference'. Will not report event: 'Warning' 'Failed' 'PipelineRun "test-pipeline-retry-run-with-timeout" failed to finish within "&Duration{Duration:12h0m0s,}"'
PASS
panic: Log in goroutine after TestReconcileWithFailingConditionChecks has completed

goroutine 210 [running]:
testing.(*common).logDepth(0xc00068c400, 0xc00067a240, 0xb2, 0x3)
	/usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:634 +0x51a
testing.(*common).log(...)
	/usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:614
testing.(*common).Logf(0xc00068c400, 0x2d6df2f, 0x2, 0xc000726ac0, 0x1, 0x1)
	/usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:649 +0x91
github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zaptest.testingWriter.Write(0x3191660, 0xc00068c400, 0x0, 0xc00086e000, 0xb3, 0x400, 0xc00000efa0, 0xc000726ab0, 0xc00072cc00)
	/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zaptest/logger.go:130 +0x120
github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore.(*ioCore).Write(0xc0001e87b0, 0x0, 0xbf56633b048c1fa0, 0x5d0c539, 0x4314120, 0x2d9c8b3, 0x27, 0xc0002244b0, 0x46, 0x1, ...)
	/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore/core.go:90 +0x1c5
github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc00057af20, 0x0, 0x0, 0x0)
	/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore/entry.go:215 +0x1e8
github.com/tektoncd/pipeline/vendor/go.uber.org/zap.(*SugaredLogger).log(0xc0001301a8, 0x0, 0x2d83fa0, 0x15, 0xc000081dd0, 0x1, 0x1, 0x0, 0x0, 0x0)
	/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/sugar.go:234 +0x143
github.com/tektoncd/pipeline/vendor/go.uber.org/zap.(*SugaredLogger).Infof(...)
	/go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/sugar.go:138
github.com/tektoncd/pipeline/pkg/reconciler.(*TimeoutSet).setTimer(0xc0000517c0, 0x3135e80, 0xc000777680, 0x34630b69302, 0xc00038fb40)
	/go/src/github.com/tektoncd/pipeline/pkg/reconciler/timeout_handler.go:281 +0x5ad
github.com/tektoncd/pipeline/pkg/reconciler.(*TimeoutSet).waitRun(0xc0000517c0, 0x3135e80, 0xc000777680, 0x34630b8a000, 0xc00092e900, 0xc00038fb40)
	/go/src/github.com/tektoncd/pipeline/pkg/reconciler/timeout_handler.go:257 +0x2db
github.com/tektoncd/pipeline/pkg/reconciler.(*TimeoutSet).WaitPipelineRun(0xc0000517c0, 0xc000777680, 0xc00092e900)
	/go/src/github.com/tektoncd/pipeline/pkg/reconciler/timeout_handler.go:243 +0xbe
created by github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun.(*Reconciler).Reconcile
	/go/src/github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun/pipelinerun.go:144 +0x17b9
FAIL	github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun	0.264s
```

If I enable `TestReconcilePropagateLabels` or `TestReconcile_InvalidPipelineRuns` I get a DATA race:

```
E0911 16:47:00.388462   80223 event.go:259] Could not construct reference to: '&v1alpha1.PipelineRun{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"test-pipeline-run-with-timeout", GenerateName:"", Namespace:"foo", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"tekton.dev/pipeline":"test-pipeline"}, Annotations:map[string]string{}, OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:v1alpha1.PipelineRunSpec{PipelineRef:v1alpha1.PipelineRef{Name:"test-pipeline", APIVersion:""}, Resources:[]v1alpha1.PipelineResourceBinding(nil), Params:[]v1alpha1.Param(nil), ServiceAccount:"test-sa", ServiceAccounts:[]v1alpha1.PipelineRunSpecServiceAccount(nil), Results:(*v1alpha1.Results)(nil), Status:"", Timeout:(*v1.Duration)(0xc0001195f0), PodTemplate:v1alpha1.PodTemplate{NodeSelector:map[string]string(nil), Tolerations:[]v1.Toleration(nil), Affinity:(*v1.Affinity)(nil), SecurityContext:(*v1.PodSecurityContext)(nil), Volumes:[]v1.Volume(nil)}, NodeSelector:map[string]string(nil), Tolerations:[]v1.Toleration(nil), Affinity:(*v1.Affinity)(nil)}, Status:v1alpha1.PipelineRunStatus{Status:v1beta1.Status{ObservedGeneration:0, Conditions:v1beta1.Conditions{apis.Condition{Type:"Succeeded", Status:"False", Severity:"", LastTransitionTime:apis.VolatileTime{Inner:v1.Time{Time:time.Time{wall:0xbf56639d17270c38, ext:73004131, loc:(*time.Location)(0x430d100)}}}, Reason:"PipelineRunTimeout", Message:"PipelineRun \"test-pipeline-run-with-timeout\" failed to finish within \"&Duration{Duration:12h0m0s,}\""}}}, Results:(*v1alpha1.Results)(nil), StartTime:(*v1.Time)(0xc0006ebb00), CompletionTime:(*v1.Time)(nil), TaskRuns:map[string]*v1alpha1.PipelineRunTaskRunStatus{}}}' due to: 'selfLink was empty, can't make reference'. Will not report event: 'Warning' 'Failed' 'PipelineRun "test-pipeline-run-with-timeout" failed to finish within "&Duration{Duration:12h0m0s,}"'
==================
WARNING: DATA RACE
Read at 0x00c0006e8b43 by goroutine 123:
  testing.(*common).logDepth()
      /usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:629 +0x132
  testing.(*common).Logf()
      /usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:614 +0x90
  testing.(*T).Logf()
      <autogenerated>:1 +0x75
  github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zaptest.testingWriter.Write()
      /go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zaptest/logger.go:130 +0x11f
  github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zaptest.(*testingWriter).Write()
      <autogenerated>:1 +0xa9
  github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore.(*ioCore).Write()
      /go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore/core.go:90 +0x1c4
  github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore.(*CheckedEntry).Write()
      /go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/zapcore/entry.go:215 +0x1e7
  github.com/tektoncd/pipeline/vendor/go.uber.org/zap.(*SugaredLogger).log()
      /go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/sugar.go:234 +0x142
  github.com/tektoncd/pipeline/pkg/reconciler.(*TimeoutSet).setTimer()
      /go/src/github.com/tektoncd/pipeline/vendor/go.uber.org/zap/sugar.go:138 +0x5ac
  github.com/tektoncd/pipeline/pkg/reconciler.(*TimeoutSet).waitRun()
      /go/src/github.com/tektoncd/pipeline/pkg/reconciler/timeout_handler.go:257 +0x2da
  github.com/tektoncd/pipeline/pkg/reconciler.(*TimeoutSet).WaitPipelineRun()
      /go/src/github.com/tektoncd/pipeline/pkg/reconciler/timeout_handler.go:243 +0xbd

Previous write at 0x00c0006e8b43 by goroutine 108:
  testing.tRunner.func1()
      /usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:856 +0x354
  testing.tRunner()
      /usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:869 +0x17f

Goroutine 123 (running) created at:
  github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun.(*Reconciler).Reconcile()
      /go/src/github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun/pipelinerun.go:144 +0x17b8
  github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun.TestReconcilePropagateLabels.func1()
      /go/src/github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun/pipelinerun_test.go:918 +0x582
  testing.tRunner()
      /usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:865 +0x163

Goroutine 108 (finished) created at:
  testing.(*T).Run()
      /usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:916 +0x65a
  testing.runTests.func1()
      /usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:1157 +0xa8
  testing.tRunner()
      /usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:865 +0x163
  testing.runTests()
      /usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:1155 +0x523
  testing.(*M).Run()
      /usr/local/Cellar/go/1.12.4/libexec/src/testing/testing.go:1072 +0x2eb
  main.main()
      _testmain.go:68 +0x222
==================
E0911 16:47:00.406244   80223 event.go:259] Could not construct reference to: '&v1alpha1.PipelineRun{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"test-pipeline-retry-run-with-timeout", GenerateName:"", Namespace:"foo", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"tekton.dev/pipeline":"test-pipeline-retry"}, Annotations:map[string]string{}, OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:v1alpha1.PipelineRunSpec{PipelineRef:v1alpha1.PipelineRef{Name:"test-pipeline-retry", APIVersion:""}, Resources:[]v1alpha1.PipelineResourceBinding(nil), Params:[]v1alpha1.Param(nil), ServiceAccount:"test-sa", ServiceAccounts:[]v1alpha1.PipelineRunSpecServiceAccount(nil), Results:(*v1alpha1.Results)(nil), Status:"", Timeout:(*v1.Duration)(0xc00024e380), PodTemplate:v1alpha1.PodTemplate{NodeSelector:map[string]string(nil), Tolerations:[]v1.Toleration(nil), Affinity:(*v1.Affinity)(nil), SecurityContext:(*v1.PodSecurityContext)(nil), Volumes:[]v1.Volume(nil)}, NodeSelector:map[string]string(nil), Tolerations:[]v1.Toleration(nil), Affinity:(*v1.Affinity)(nil)}, Status:v1alpha1.PipelineRunStatus{Status:v1beta1.Status{ObservedGeneration:0, Conditions:v1beta1.Conditions{apis.Condition{Type:"Succeeded", Status:"False", Severity:"", LastTransitionTime:apis.VolatileTime{Inner:v1.Time{Time:time.Time{wall:0xbf56639d183697d8, ext:90799949, loc:(*time.Location)(0x430d100)}}}, Reason:"PipelineRunTimeout", Message:"PipelineRun \"test-pipeline-retry-run-with-timeout\" failed to finish within \"&Duration{Duration:12h0m0s,}\""}}}, Results:(*v1alpha1.Results)(nil), StartTime:(*v1.Time)(0xc000437c80), CompletionTime:(*v1.Time)(nil), TaskRuns:map[string]*v1alpha1.PipelineRunTaskRunStatus{"hello-world-1":(*v1alpha1.PipelineRunTaskRunStatus)(0xc000437ca0)}}}' due to: 'selfLink was empty, can't make reference'. Will not report event: 'Warning' 'Failed' 'PipelineRun "test-pipeline-retry-run-with-timeout" failed to finish within "&Duration{Duration:12h0m0s,}"'
E0911 16:47:00.408880   80223 event.go:259] Could not construct reference to: '&v1alpha1.PipelineRun{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"test-pipeline-retry-run-with-timeout", GenerateName:"", Namespace:"foo", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"tekton.dev/pipeline":"test-pipeline-retry"}, Annotations:map[string]string{}, OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:v1alpha1.PipelineRunSpec{PipelineRef:v1alpha1.PipelineRef{Name:"test-pipeline-retry", APIVersion:""}, Resources:[]v1alpha1.PipelineResourceBinding(nil), Params:[]v1alpha1.Param(nil), ServiceAccount:"test-sa", ServiceAccounts:[]v1alpha1.PipelineRunSpecServiceAccount(nil), Results:(*v1alpha1.Results)(nil), Status:"", Timeout:(*v1.Duration)(0xc0001184e0), PodTemplate:v1alpha1.PodTemplate{NodeSelector:map[string]string(nil), Tolerations:[]v1.Toleration(nil), Affinity:(*v1.Affinity)(nil), SecurityContext:(*v1.PodSecurityContext)(nil), Volumes:[]v1.Volume(nil)}, NodeSelector:map[string]string(nil), Tolerations:[]v1.Toleration(nil), Affinity:(*v1.Affinity)(nil)}, Status:v1alpha1.PipelineRunStatus{Status:v1beta1.Status{ObservedGeneration:0, Conditions:v1beta1.Conditions{apis.Condition{Type:"Succeeded", Status:"False", Severity:"", LastTransitionTime:apis.VolatileTime{Inner:v1.Time{Time:time.Time{wall:0xbf56639d185ee058, ext:93440706, loc:(*time.Location)(0x430d100)}}}, Reason:"PipelineRunTimeout", Message:"PipelineRun \"test-pipeline-retry-run-with-timeout\" failed to finish within \"&Duration{Duration:12h0m0s,}\""}}}, Results:(*v1alpha1.Results)(nil), StartTime:(*v1.Time)(0xc00000fb40), CompletionTime:(*v1.Time)(nil), TaskRuns:map[string]*v1alpha1.PipelineRunTaskRunStatus{"hello-world-1":(*v1alpha1.PipelineRunTaskRunStatus)(0xc00000fb60)}}}' due to: 'selfLink was empty, can't make reference'. Will not report event: 'Warning' 'Failed' 'PipelineRun "test-pipeline-retry-run-with-timeout" failed to finish within "&Duration{Duration:12h0m0s,}"'
FAIL
FAIL	github.com/tektoncd/pipeline/pkg/reconciler/pipelinerun	0.263s
```

So perhaps there are two different issues happening here?

--- a/pkg/apis/pipeline/v1alpha1/pipelinerun_types.go
+++ b/pkg/apis/pipeline/v1alpha1/pipelinerun_types.go
@@ -246,7 +246,8 @@ func (pr *PipelineRun) IsCancelled() bool {
 
 // GetRunKey return the pipelinerun key for timeout handler map
 func (pr *PipelineRun) GetRunKey() string {
-	return fmt.Sprintf("%s/%s/%s", pipelineRunControllerName, pr.Namespace, pr.Name)
+	// The address of the pointer is a threadsafe unique identifier for the pipelinerun
+	return fmt.Sprintf("%s/%p", pipelineRunControllerName, pr)
 }
 
 // IsTimedOut returns true if a pipelinerun has exceeded its spec.Timeout based on its status.Timeout

--- a/pkg/apis/pipeline/v1alpha1/pipelinerun_types_test.go
+++ b/pkg/apis/pipeline/v1alpha1/pipelinerun_types_test.go
@@ -17,6 +17,7 @@ limitations under the License.
 package v1alpha1_test
 
 import (
+	"fmt"
 	"testing"
 	"time"
 
@@ -130,13 +131,8 @@ func TestPipelineRunIsCancelled(t *testing.T) {
 }
 
 func TestPipelineRunKey(t *testing.T) {
-	pr := &v1alpha1.PipelineRun{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      "prunname",
-			Namespace: "testns",
-		},
-	}
-	expectedKey := "PipelineRun/testns/prunname"
+	pr := tb.PipelineRun("prunname", "testns")
+	expectedKey := fmt.Sprintf("PipelineRun/%p", pr)
 	if pr.GetRunKey() != expectedKey {
 		t.Fatalf("Expected taskrun key to be %s but got %s", expectedKey, pr.GetRunKey())
 	}

--- a/pkg/apis/pipeline/v1alpha1/taskrun_types.go
+++ b/pkg/apis/pipeline/v1alpha1/taskrun_types.go
@@ -291,5 +291,6 @@ func (tr *TaskRun) IsCancelled() bool {
 
 // GetRunKey return the taskrun key for timeout handler map
 func (tr *TaskRun) GetRunKey() string {
-	return fmt.Sprintf("%s/%s/%s", "TaskRun", tr.Namespace, tr.Name)
+	// The address of the pointer is a threadsafe unique identifier for the taskrun
+	return fmt.Sprintf("%s/%p", "TaskRun", tr)
 }

--- a/pkg/apis/pipeline/v1alpha1/taskrun_types_test.go
+++ b/pkg/apis/pipeline/v1alpha1/taskrun_types_test.go
@@ -17,6 +17,7 @@ limitations under the License.
 package v1alpha1_test
 
 import (
+	"fmt"
 	"testing"
 	"time"
 
@@ -112,8 +113,8 @@ func TestTaskRunIsCancelled(t *testing.T) {
 }
 
 func TestTaskRunKey(t *testing.T) {
-	tr := tb.TaskRun("taskrunname", "testns")
-	expectedKey := "TaskRun/testns/taskrunname"
+	tr := tb.TaskRun("taskrunname", "")
+	expectedKey := fmt.Sprintf("TaskRun/%p", tr)
 	if tr.GetRunKey() != expectedKey {
 		t.Fatalf("Expected taskrun key to be %s but got %s", expectedKey, tr.GetRunKey())
 	}

--- a/pkg/reconciler/timeout_handler.go
+++ b/pkg/reconciler/timeout_handler.go
@@ -170,7 +170,6 @@ func backoffDuration(count uint, jf jitterFunc) time.Duration {
 func (t *TimeoutSet) checkPipelineRunTimeouts(namespace string, pipelineclientset clientset.Interface) {
 	pipelineRuns, err := pipelineclientset.TektonV1alpha1().PipelineRuns(namespace).List(metav1.ListOptions{})
 	if err != nil {
-		t.logger.Errorf("Can't get pipelinerun list in namespace %s: %s", namespace, err)
 		return
 	}
 	for _, pipelineRun := range pipelineRuns.Items {
@@ -189,7 +188,6 @@ func (t *TimeoutSet) checkPipelineRunTimeouts(namespace string, pipelineclientse
 func (t *TimeoutSet) CheckTimeouts(kubeclientset kubernetes.Interface, pipelineclientset clientset.Interface) {
 	namespaces, err := kubeclientset.CoreV1().Namespaces().List(metav1.ListOptions{})
 	if err != nil {
-		t.logger.Errorf("Can't get namespaces list: %s", err)
 		return
 	}
 	for _, namespace := range namespaces.Items {
@@ -203,7 +201,6 @@ func (t *TimeoutSet) CheckTimeouts(kubeclientset kubernetes.Interface, pipelinec
 func (t *TimeoutSet) checkTaskRunTimeouts(namespace string, pipelineclientset clientset.Interface) {
 	taskruns, err := pipelineclientset.TektonV1alpha1().TaskRuns(namespace).List(metav1.ListOptions{})
 	if err != nil {
-		t.logger.Errorf("Can't get taskrun list in namespace %s: %s", namespace, err)
 		return
 	}
 	for _, taskrun := range taskruns.Items {
@@ -245,14 +242,12 @@ func (t *TimeoutSet) WaitPipelineRun(pr *v1alpha1.PipelineRun, startTime *metav1
 
 func (t *TimeoutSet) waitRun(runObj StatusKey, timeout time.Duration, startTime *metav1.Time, callback func(interface{})) {
 	if startTime == nil {
-		t.logger.Errorf("startTime must be specified in order for a timeout to be calculated accurately for %s", runObj.GetRunKey())
 		return
 	}
 	if callback == nil {
 		callback = defaultFunc
 	}
 	runtime := time.Since(startTime.Time)
-	t.logger.Infof("About to start timeout timer for %s. started at %s, timeout is %s, running for %s", runObj.GetRunKey(), startTime.Time, timeout, runtime)
 	defer t.Release(runObj)
 	t.setTimer(runObj, timeout-runtime, callback)
 }
@@ -267,7 +262,6 @@ func (t *TimeoutSet) waitRun(runObj StatusKey, timeout time.Duration, startTime 
 func (t *TimeoutSet) SetTaskRunTimer(tr *v1alpha1.TaskRun, d time.Duration) {
 	callback := t.taskRunCallbackFunc
 	if callback == nil {
-		t.logger.Errorf("attempted to set a timer for %q but no task run callback has been assigned", tr.Name)
 		return
 	}
 	t.setTimer(tr, d, callback)
@@ -275,16 +269,12 @@ func (t *TimeoutSet) SetTaskRunTimer(tr *v1alpha1.TaskRun, d time.Duration) {
 
 func (t *TimeoutSet) setTimer(runObj StatusKey, timeout time.Duration, callback func(interface{})) {
 	finished := t.getOrCreateFinishedChan(runObj)
-	started := time.Now()
 	select {
 	case <-t.stopCh:
-		t.logger.Infof("stopping timer for %q", runObj.GetRunKey())
 		return
 	case <-finished:
-		t.logger.Infof("%q finished, stopping timer", runObj.GetRunKey())
 		return
 	case <-time.After(timeout):
-		t.logger.Infof("timer for %q has activated after %s", runObj.GetRunKey(), time.Since(started).String())
 		callback(runObj)
 	}
 }

--- a/test/controller.go
+++ b/test/controller.go
@@ -35,21 +35,11 @@ import (
 	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
 	fakepipelineclientset "github.com/tektoncd/pipeline/pkg/client/clientset/versioned/fake"
 	informersv1alpha1 "github.com/tektoncd/pipeline/pkg/client/informers/externalversions/pipeline/v1alpha1"
-	"go.uber.org/zap/zaptest/observer"
 	corev1 "k8s.io/api/core/v1"
 	coreinformers "k8s.io/client-go/informers/core/v1"
 	fakekubeclientset "k8s.io/client-go/kubernetes/fake"
 	"knative.dev/pkg/controller"
 )
-
-// GetLogMessages returns a list of all string logs in logs.
-func GetLogMessages(logs *observer.ObservedLogs) []string {
-	messages := []string{}
-	for _, l := range logs.All() {
-		messages = append(messages, l.Message)
-	}
-	return messages
-}
 
 // Data represents the desired state of the system (i.e. existing resources) to seed controllers
 // with.
@@ -86,7 +76,6 @@ type Informers struct {
 // TestAssets holds references to the controller, logs, clients, and informers.
 type TestAssets struct {
 	Controller *controller.Impl
-	Logs       *observer.ObservedLogs
 	Clients    Clients
 }
 

